// autocomplete/db.js
var storage = require('./storage');
var node = require('./node');
var tokenize = require('./tokenize');
var Index = require('./db-index');

var tree;
var dbIndex;

function load() {
  tree = storage.getTree();
  dbIndex = new Index(storage.getTokenIndex());
}

function save() {
  storage.storeTree(tree);
  storage.storeTokenIndex(dbIndex.hash);
  storage.setUpdatedAt(new Date().toString());
}

function truncate() {
  tree = {};
  dbIndex = new Index();
}

function index(string) {
  var tokens = tokenize(string);
  if (!tokens) return;

  tokens.forEach(function(token) {
    token = token.toLowerCase();

    dbIndex.add(token, string)

    indexToken(token);
  });
}

function indexToken(word) {
  var currentLevel = node(tree);

  for (var i in word) {
    currentLevel = node(currentLevel, word[i]);
  }

  currentLevel.setIsWord();
}

function search(term, limit) {
  term || (term = '');
  if (term == '') return [];

  term = term.toLowerCase();

  var subtree = tree;

  for (var i in term) {
    subtree = node(subtree, term[i]);
  }

  var queue = [subtree];
  var results = [];

  while(queue.length > 0) {
    var currentNode = node(queue.shift());

    if (currentNode.isWord) Array.prototype.push.apply(results, dbIndex.lookup(currentNode.getWord()));
    if (limit && results.length >= limit) break;

    currentNode.eachChild(function(childNode) {
      queue.push(childNode);
    });
  }

  return results.slice(0, limit);
}

function isEmpty() {
  return Object.keys(tree).length == 0;
}


truncate();


module.exports = {
  load: load,
  save: save,
  index: index,
  search: search,
  isEmpty: isEmpty,
  getUpdatedAt: storage.getUpdatedAt,
  truncate: truncate
}
